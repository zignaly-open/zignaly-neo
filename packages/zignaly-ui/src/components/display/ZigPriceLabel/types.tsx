import { Typography, TypographyProps } from "@mui/material";

export type ZigPriceLabelProps = Partial<TypographyProps> & {
  value: string | number;
  coin?: string;
  precision?: number;
  exact?: boolean;
  prefix?: string | JSX.Element;
  shorten?: boolean;
  showTooltip?: boolean;
  usd?: boolean;
  coinProps?: Partial<typeof Typography>;
  component?: string;
  alwaysShowSign?: boolean;
  id?: string;
  showApproximate?: boolean;
  showCoinIcon?: boolean;
  label?: string;
};
