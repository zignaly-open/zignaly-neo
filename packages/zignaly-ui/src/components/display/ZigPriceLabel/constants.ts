// TODO: better way
export const coinPrecisions: Record<string, number> = {
  USDT: 2,
  USDC: 2,
  BUSD: 2,
  BTC: 8,
};
