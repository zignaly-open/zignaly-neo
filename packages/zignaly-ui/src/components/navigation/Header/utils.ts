// TODO: Move this for config or theme folder.
export const headerConfig = {
  height: "52px",
};
