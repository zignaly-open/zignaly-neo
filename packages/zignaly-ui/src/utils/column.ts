import styled from "styled-components";

export const Column = styled.div<{
  gap?: number;
  justifyContent?: "center" | "end" | "start" | "space-between";
  alignItems?: "center" | "end" | "start" | "space-between";
  textAlign?: "end" | "center" | "start";
  height?: string;
  flex?: number;
  width?: string;
}>`
  display: flex;
  flex-direction: column;
  flex: ${(props) => props.flex};
  gap: ${(props) => props.gap}px;
  justify-content: ${(props) => props.justifyContent};
  align-items: ${(props) => props.alignItems};
  text-align: ${(props) => `${props.textAlign}`};
  height: ${(props) => `${props.height}`};
  width: ${(props) => `${props.width}`};
`;
