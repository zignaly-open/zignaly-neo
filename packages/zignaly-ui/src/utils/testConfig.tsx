import { render } from "@testing-library/react";
import { ThemeProvider as ThemeProviderMui } from "@mui/material";
import { ThemeProvider, ThemeProviderMui as ThemeInheritorMui } from "../index";
import React from "react";
import { getZignalyThemeExport } from "../theme";
import mediaQuery from "css-mediaquery";
import { JestI18nextDecorator } from "./dummyI18nextProvider";

export const renderWithProvidersUi = (ui: JSX.Element, renderOptions = {}) => {
  const { mui, legacyStyledComponentsDoNotUse } = getZignalyThemeExport();
  return render(ui, {
    wrapper: ({ children }: { children: JSX.Element }) => (
      <JestI18nextDecorator>
        <ThemeProvider theme={legacyStyledComponentsDoNotUse}>
          <ThemeInheritorMui theme={mui}>
            <ThemeProviderMui theme={mui}>{children}</ThemeProviderMui>
          </ThemeInheritorMui>
        </ThemeProvider>
      </JestI18nextDecorator>
    ),
    ...renderOptions,
  });
};

const createMatchMedia = (width: number) => {
  return (query: string) => {
    return {
      matches: mediaQuery.match(query, { width }),
      media: "",
      addListener: () => {},
      removeListener: () => {},
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      dispatchEvent: () => true,
    };
  };
};

export const resizeScreenSize = (width: number) => {
  window.matchMedia = createMatchMedia(width);
};
