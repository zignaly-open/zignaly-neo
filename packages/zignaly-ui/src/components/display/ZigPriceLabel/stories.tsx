import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import ZigPriceLabel from "./index";

export default {
  title: "Display/ZigPriceLabel",
  component: ZigPriceLabel,
  argTypes: {
    value: {
      control: "number",
      default: 50,
    },
    coin: {
      control: "text",
      default: "USDT",
    },
    fiat: {
      control: "boolean",
      default: false,
    },
    symbol: {
      control: "text",
      default: "$",
    },
  },
} as ComponentMeta<typeof ZigPriceLabel>;

const Template: ComponentStory<typeof ZigPriceLabel> = (args) => <ZigPriceLabel {...args} />;

export const USDT = Template.bind({});
USDT.args = {
  value: 50,
  coin: "USDT",
};

export const ETH = Template.bind({});
ETH.args = {
  value: Math.PI,
  coin: "ETH",
};

export const Dollars = Template.bind({});
Dollars.args = {
  value: Math.PI,
  usd: true,
};

export const Styled = Template.bind({});
Styled.args = {
  fontWeight: "bold",
  color: "greenGraph", // todo: rename those stupid colors to green and red
  value: 500000000,
  variant: "h1",
  coin: "BTC",
  sx: {
    textDecoration: "underline",
    background: "rgba(0,0,0,.5)",
    border: "3px dashed green",
    display: "inline-block",
    padding: "20px",
  },
  coinProps: {
    fontWeight: 800,
    variant: "h1",
    sx: {
      display: "inline-block",
      color: "red",
    },
  },
};

export const Precision = Template.bind({});
Precision.args = {
  precision: 3,
  coin: "BTC",
  value: Math.PI,
};
