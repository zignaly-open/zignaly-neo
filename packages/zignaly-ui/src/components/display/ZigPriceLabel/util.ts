import { coinPrecisions } from "./constants";

export function getPrecisionForCoin(coin: string, value: string | number): number {
  return (+value === 0 && 2) || coinPrecisions[coin] || 8;
}
