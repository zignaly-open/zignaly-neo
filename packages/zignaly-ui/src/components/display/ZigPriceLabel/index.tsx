import React from "react";
import { NumericFormat, numericFormatter } from "react-number-format";
import { ZigPriceLabelProps } from "./types";
import { getPrecisionForCoin, shortenNumber } from "./util";
import ZigTypography from "../ZigTypography";
import { Variant } from "@mui/material/styles/createTypography";
import { Tooltip } from "@mui/material";

const ZigPriceLabel: React.FC<ZigPriceLabelProps> = ({
  value = 0,
  coin,
  precision,
  shorten,
  prefix,
  exact,
  usd,
  coinProps,
  showTooltip = !usd,
  alwaysShowSign = false,
  ...otherProps
}) => {
  const withDefaultPropsCoin = {
    variant: "body2" as Variant,
    color: "neutral200",
    fontWeight: "regular",
    ...(coinProps || ""),
  };

  const withDefaultProps = {
    color: "almostWhite",
    variant: "body1" as Variant,
    fontWeight: "500",
    ...(otherProps || ""),
  };

  const {
    value: shortened,
    precision: shortenedPrecision,
    suffix: shortenSuffix,
  } = shortenNumber(+value);

  const content = (
    <ZigTypography
      {...withDefaultProps}
      sx={{ whiteSpace: "nowrap", ...(withDefaultProps?.sx || {}) }}
    >
      {!!prefix && <>{prefix}</>}
      {+value >= 0 ? alwaysShowSign ? "+" : "" : <>&ndash;</>}
      {usd && "$"}
      <NumericFormat
        value={Math.abs(shorten ? shortened : +value)}
        renderText={(v) => v}
        displayType={"text"}
        thousandSeparator={true}
        decimalScale={
          exact
            ? undefined
            : shorten
            ? shortenedPrecision
            : precision || getPrecisionForCoin(coin || "USDT", value)
        }
      />

      {shorten ? shortenSuffix : ""}

      {coin && (
        <>
          {" "}
          <ZigTypography {...withDefaultPropsCoin}>{coin}</ZigTypography>
        </>
      )}
    </ZigTypography>
  );

  return showTooltip || shorten ? (
    <Tooltip
      title={
        numericFormatter(value?.toString() ?? "", {
          thousandSeparator: true,
          displayType: "text",
        }) +
        " " +
        coin
      }
    >
      {content}
    </Tooltip>
  ) : (
    content
  );
};

export default React.memo(ZigPriceLabel);

export const ZigTablePriceLabel = (props: ZigPriceLabelProps) => (
  <ZigPriceLabel color="neutral100" coinProps={{ color: "neutral400" }} {...props} />
);
