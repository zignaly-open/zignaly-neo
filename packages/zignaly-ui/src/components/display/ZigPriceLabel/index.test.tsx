import React from "react";
import ZigPriceLabel, { ZigTablePriceLabel } from "./index";
import { renderWithProvidersUi } from "../../../utils/testConfig";

describe("components/display/ZigPriceLabel", () => {
  describe("ZigPriceLabel", () => {
    it("should render without crashing", async () => {
      const { container } = renderWithProvidersUi(
        <ZigPriceLabel id={"test-id"} value={500} coin={"usdt"} />,
      );

      const priceLabel = container.querySelector("#test-id") as Element;
      expect(priceLabel).toBeVisible();
      const priceLabelStyles = getComputedStyle(priceLabel);
      expect(priceLabel).toMatchSnapshot();
      expect(priceLabelStyles).toMatchSnapshot();
    });
  });

  describe("ZigTablePriceLabel", () => {
    it("should render without crashing", async () => {
      const { container } = renderWithProvidersUi(
        <ZigTablePriceLabel id={"test-id"} value={500} coin={"btc"} />,
      );

      const priceLabel = container.querySelector("#test-id") as Element;
      expect(priceLabel).toBeVisible();
      const priceLabelStyles = getComputedStyle(priceLabel);
      expect(priceLabel).toMatchSnapshot();
      expect(priceLabelStyles).toMatchSnapshot();
    });
  });
});
