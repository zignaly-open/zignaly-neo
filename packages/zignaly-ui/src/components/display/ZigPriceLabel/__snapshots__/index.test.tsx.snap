// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/display/ZigPriceLabel ZigPriceLabel rendering renders 1`] = `
<span
  aria-label="500 usdt"
  class="MuiTypography-root MuiTypography-body1 css-1c6zzem-MuiTypography-root"
  data-mui-internal-clone-element="true"
  id="test-id"
>
  500
   
  <span
    class="MuiTypography-root MuiTypography-body2 test-id-coin css-1e0f0rf-MuiTypography-root"
  >
    usdt
  </span>
</span>
`;

exports[`components/display/ZigPriceLabel ZigPriceLabel rendering renders 2`] = `
CSSStyleDeclaration {
  "0": "margin",
  "1": "font-family",
  "10": "justify-content",
  "11": "visibility",
  "2": "font-weight",
  "3": "font-size",
  "4": "line-height",
  "5": "letter-spacing",
  "6": "color",
  "7": "white-space",
  "8": "display",
  "9": "align-items",
  "_importants": {
    "align-items": "",
    "color": "",
    "display": "",
    "font-family": "",
    "font-size": "",
    "font-weight": "",
    "justify-content": "",
    "letter-spacing": "",
    "line-height": "",
    "margin": "",
    "visibility": undefined,
    "white-space": "",
  },
  "_length": 12,
  "_onChange": [Function],
  "_values": {
    "align-items": "center",
    "color": "rgb(243, 244, 246)",
    "display": "inline-flex",
    "font-family": "Avenir Next,Roboto,Helvetica,Arial,sans-serif",
    "font-size": "15px",
    "font-weight": "500",
    "justify-content": "center",
    "letter-spacing": "0.55px",
    "line-height": "24px",
    "margin": "0px",
    "margin-bottom": "0px",
    "margin-left": "0px",
    "margin-right": "0px",
    "margin-top": "0px",
    "visibility": "visible",
    "white-space": "nowrap",
  },
}
`;

exports[`components/display/ZigPriceLabel ZigTablePriceLabel rendering renders 1`] = `
<span
  aria-label="500 btc"
  class="MuiTypography-root MuiTypography-body1 css-10ef9gy-MuiTypography-root"
  data-mui-internal-clone-element="true"
  id="test-id"
>
  500
   
  <span
    class="MuiTypography-root MuiTypography-body2 test-id-coin css-1k4dd5e-MuiTypography-root"
  >
    btc
  </span>
</span>
`;

exports[`components/display/ZigPriceLabel ZigTablePriceLabel rendering renders 2`] = `
CSSStyleDeclaration {
  "0": "margin",
  "1": "font-family",
  "10": "justify-content",
  "11": "visibility",
  "2": "font-weight",
  "3": "font-size",
  "4": "line-height",
  "5": "letter-spacing",
  "6": "color",
  "7": "white-space",
  "8": "display",
  "9": "align-items",
  "_importants": {
    "align-items": "",
    "color": "",
    "display": "",
    "font-family": "",
    "font-size": "",
    "font-weight": "",
    "justify-content": "",
    "letter-spacing": "",
    "line-height": "",
    "margin": "",
    "visibility": undefined,
    "white-space": "",
  },
  "_length": 12,
  "_onChange": [Function],
  "_values": {
    "align-items": "center",
    "color": "rgb(193, 193, 200)",
    "display": "inline-flex",
    "font-family": "Avenir Next,Roboto,Helvetica,Arial,sans-serif",
    "font-size": "15px",
    "font-weight": "500",
    "justify-content": "center",
    "letter-spacing": "0.55px",
    "line-height": "24px",
    "margin": "0px",
    "margin-bottom": "0px",
    "margin-left": "0px",
    "margin-right": "0px",
    "margin-top": "0px",
    "visibility": "visible",
    "white-space": "nowrap",
  },
}
`;
